### ISSUE TITLE
[Snapcast] Last seconds of songs are cut when streaming to Snapcast clients

### ISSUE THREAD

**Issue by SBado** [music-assistant/support#1962]
### What version of Music Assistant has the issue?

2.0.0b106

### What version of the Home Assistant Integration have you got installed?

2024.2.0

### Have you tried everything in the Troubleshooting FAQ and reviewed the Open and Closed Issues and Discussions to resolve this yourself?

- [X] Yes

### The problem

As per title, when streaming to Snapcast clients, every songs stops playing ~10 seconds before the end of the track. By "stops playing" I mean that the web UI shows the track still playing, but no sound is reproduced.

### How to reproduce

Play a song using a Snapcast player.

### Music Providers

Filesystem (local disk)

### Player Providers

Snapcast

### Additional information

I tried to debug the problem and took a look at the Snapcast provider code, and this is what I found:

As soon as execution hits line 300 of file `music_assistant/server/providers/snapcast__init.py__`, function `play_media`, Snapcast server stop sending audio to all clients. This line is reached when all pcm chunks are sent to the Snapcast stream (the tcp socket automatically created by Music Assistant with name `MusicAssistant--{port}`). The problem is that, by this time, Snapcast still has some bytes left to send to its clients, but whent it sees the stream being destroyed it stops sending them.

To prove this, I added a small sleep of 60 seconds before the call to `self._snapserver.stream_remove_stream` and indeed the problem went away.

### What version of Home Assistant Core are your running

2024.2.3

### What type of installation are you running?

Home Assistant Container

### On what type of hardware are you running?

Alternative

**Comment by marcelveldt** [comment]
hmm annoying that there is no such thing as an auto close option for those streams. I'll add the sleep if I can not come up with something better,.

**Comment by marcelveldt** [comment]
BTW strictly taken its just the end of the queue that is cut off and not every song

**Comment by SBado** [comment]
Oh, I'm sorry, I didn't try with more than one song at a time. You are absolutely right, it's just the last song in the queue being affected. That's probably the reason why I didn't find anyone else having this problem...
Anyway, thanks for looking into it!

**Comment by OzGav** [comment]
b107 has been released

**Comment by whc2001** [comment]
This seems still happens occationally with the latest dev (last few seconds are cut off, while the progress bar goes beyond the track length with silence), however after playing a few tracks it seems to be normal again. Not sure how to trigger it consistently...

**Comment by janstadt** [comment]
Im not sure if this is the same issue, but I am experiencing audio getting cut short on announcements from MASS/HASS. 

MASS Logs:
```
2025-01-28 00:26:49.661 INFO (MainThread) [music_assistant.players] Playback announcement to player The Kitchen (Announcements) (with pre-announce: True): http://192.168.1.68:8123/api/tts_proxy/eEhvao0lFQcHIGac4SeYyQ.mp3
2025-01-28 00:26:49.662 DEBUG (MainThread) [music_assistant.players] Announcement to player The Kitchen (Announcements) - setting temporary volume (75)...
2025-01-28 00:26:49.662 DEBUG (MainThread) [music_assistant.players] Handling command cmd_volume_set for player The Kitchen (Announcements)
2025-01-28 00:26:49.682 DEBUG (MainThread) [music_assistant.players] Announcement to player The Kitchen (Announcements) - playing the announcement on the player...
2025-01-28 00:26:49.682 DEBUG (MainThread) [music_assistant.players] Handling command play_media for player The Kitchen (Announcements)
2025-01-28 00:26:49.710 DEBUG (MainThread) [music_assistant.players] Waiting for player The Kitchen (Announcements) to reach state playing
2025-01-28 00:26:49.710 DEBUG (MainThread) [music_assistant.audio] Generated ffmpeg params for player ma_b827ebe4f50b: ['alimiter=limit=-2dB:level=false:asc=true']
2025-01-28 00:26:49.840 DEBUG (MainThread) [music_assistant.players] Player The Kitchen (Announcements) reached state playing within 0.11 seconds
2025-01-28 00:26:50.392 DEBUG (MainThread) [music_assistant.players] Waiting for player The Kitchen (Announcements) to reach state idle
2025-01-28 00:26:50.571 INFO (MainThread) [snapcast.control.server] stream MusicAssistant--4971 not found, synchronize
2025-01-28 00:26:50.595 DEBUG (MainThread) [music_assistant.players] Player The Kitchen (Announcements) reached state idle too soon (0.2 vs 6.649796 seconds) - add fallback sleep...
2025-01-28 00:26:57.046 DEBUG (MainThread) [music_assistant.players] Announcement to player The Kitchen (Announcements) - restore previous state...
2025-01-28 00:26:57.047 DEBUG (MainThread) [music_assistant.players] Handling command cmd_volume_set for player The Kitchen (Announcements)


```
[eEhvao0lFQcHIGac4SeYyQ.mp3.zip](https://github.com/user-attachments/files/18566287/eEhvao0lFQcHIGac4SeYyQ.mp3.zip)

It got to "Play text on any" and then it is cut off.

**Comment by marcelveldt** [comment]
Looks like this issue was closed too early before any verification that it was solved.
I have reopened it so I can have another look although this could simply be tricky due to how snapcast works internally.

**Comment by whc2001** [comment]
I see recently there is an update on beta channel about the FFmpeg update, I'll check the current behaviour later ASAP, but since it appears randomly it's not particularly easy to test

**Comment by marcelveldt** [comment]
> I see recently there is an update on beta channel about the FFmpeg update, I'll check the current behaviour later ASAP, but since it appears randomly it's not particularly easy to test

Yes all testing should be done in the beta version because that is miles ahead of stable currently with both features and fixes, including the bump of ffmpeg which solved a bunch of vague issues due to a nasty bug in ffmpeg 6.1

**Comment by OzGav** [comment]
@whc2001 can you provide an update

**Comment by whc2001** [comment]
> [@whc2001](https://github.com/whc2001) can you provide an update

I tested with Snap.net client on Win10 and seems like things are fine for now (apart from that the seekbar often behaves erroneous), but I might need to further confirm that using the ESP32 snapclient which I'll be using primarily

**Comment by whc2001** [comment]
Update: tested with ESP32 snapclient, for now everything seems fine. If it happens again I'll report the  logs

**Comment by whc2001** [comment]
Update: seems like the reverse of the forementioned seekbar problem also happens randomly when seeking. The song will always play from the beginning, but the seekbar behaves as if the seek is correct. In that case if the seekbar is near finishing, it will play past the end.

https://github.com/user-attachments/assets/090fcbb4-d98f-4a3d-b61e-c16588c3f926

Relavent logs: [log.txt](https://github.com/user-attachments/files/18626952/log.txt)

**Comment by OzGav** [comment]
Ok if that is a different problem then raise a new issue. If the OP is fixed then close this.

**Comment by whc2001** [comment]
OK sorry but I really can't get my head around this. I tried a lot of things by modifying the Snapcast provider code but they are not related to the actual problem. I think Snapcast and MA won't work well together from the start by design (one expects continuous audio stream without ability to tell the real progress, the other only provides media URL and needs an actual player, in this case FFmpeg). I decided to ditch Snapcast completely and seek for something else (maybe Squeezelite on ESP32-S3). I am not sure whether the original problem is fixed but I can't really test stably it anymore. Thanks for your hard work.

**Comment by OzGav** [comment]
I can confirm the behaviour that seeking causes the track to restart but the track indicator moves to the selected position. @marcelveldt do you want to followup on that here or open a new issue?

**Comment by OzGav** [comment]
Beta 24 has a fix for the track restart. Please confirm this is resolved

**Comment by OzGav** [comment]
I confirm that seek has been fixed with Snapcast in the latest beta
